File: src/Sample/Flutter.Net/AppBarIconThemeDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Material;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/app_bar_icon_theme_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class AppBarIconThemeDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AppBarIconThemeDemoPageState();
    }
}

internal sealed class AppBarIconThemeDemoPageState : State
{
    private static readonly Color AppBarBackground = Color.Parse("#FF1E3A5F");
    private static readonly Color ForegroundWhite = Colors.White;
    private static readonly Color ForegroundMint = Color.Parse("#FFB8FFF1");
    private static readonly Color ThemeIconColor = Color.Parse("#FF2A9D8F");
    private static readonly Color ThemeActionsIconColor = Color.Parse("#FFF4A261");
    private static readonly Color WidgetIconColor = Color.Parse("#FF8E44AD");
    private static readonly Color WidgetActionsIconColor = Color.Parse("#FFE63946");

    private Color _foregroundColor = ForegroundWhite;
    private bool _themeIconEnabled = true;
    private bool _themeActionsIconEnabled;
    private bool _widgetIconOverrideEnabled;
    private bool _widgetActionsIconOverrideEnabled;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AppBar iconTheme/actionsIconTheme", fontSize: 20, color: Colors.Black),
                new Text(
                    "Runtime probe for icon-theme precedence: widget override > appBarTheme > iconTheme fallback > foreground fallback.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("fg white", () => SetForeground(ForegroundWhite), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("fg mint", () => SetForeground(ForegroundMint), width: 92, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme icon off", () => SetThemeIconEnabled(false), width: 108, colorHex: "#FFE9F5EC"),
                        BuildButton("theme icon on", () => SetThemeIconEnabled(true), width: 108, colorHex: "#FFE9F5EC"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme actions off", () => SetThemeActionsIconEnabled(false), width: 118, colorHex: "#FFF3E8D8"),
                        BuildButton("theme actions on", () => SetThemeActionsIconEnabled(true), width: 118, colorHex: "#FFF3E8D8"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget icon off", () => SetWidgetIconOverrideEnabled(false), width: 114, colorHex: "#FFE4ECF7"),
                        BuildButton("widget icon on", () => SetWidgetIconOverrideEnabled(true), width: 114, colorHex: "#FFE4ECF7"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget actions off", () => SetWidgetActionsIconOverrideEnabled(false), width: 126, colorHex: "#FFF5E8ED"),
                        BuildButton("widget actions on", () => SetWidgetActionsIconOverrideEnabled(true), width: 126, colorHex: "#FFF5E8ED"),
                    ]),
                new Text(
                    $"fg={ColorLabel(_foregroundColor)}, themeIcon={OnOff(_themeIconEnabled)}, themeActions={OnOff(_themeActionsIconEnabled)}, widgetIcon={OnOff(_widgetIconOverrideEnabled)}, widgetActions={OnOff(_widgetActionsIconOverrideEnabled)}, expectedLeading={ColorLabel(ResolveExpectedLeadingColor())}, expectedActions={ColorLabel(ResolveExpectedActionsColor())}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 320,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text("Themed app bar preview", fontSize: 11, color: Colors.DimGray),
                            BuildThemedPreview(),
                            new Text("Default app bar reference (no icon theme overrides)", fontSize: 11, color: Colors.DimGray),
                            BuildDefaultReferencePreview(),
                        ])),
            ]);
    }

    private Widget BuildThemedPreview()
    {
        return new Theme(
            data: ThemeData.Light with
            {
                AppBarTheme = new AppBarThemeData(
                    BackgroundColor: AppBarBackground,
                    ForegroundColor: _foregroundColor,
                    IconTheme: _themeIconEnabled ? new IconThemeData(Color: ThemeIconColor, Size: 18) : null,
                    ActionsIconTheme: _themeActionsIconEnabled ? new IconThemeData(Color: ThemeActionsIconColor, Size: 16) : null),
            },
            child: new AppBar(
                titleText: "Theme icon chain",
                iconTheme: _widgetIconOverrideEnabled ? new IconThemeData(Color: WidgetIconColor, Size: 20) : null,
                actionsIconTheme: _widgetActionsIconOverrideEnabled ? new IconThemeData(Color: WidgetActionsIconColor, Size: 22) : null,
                leading: new IconThemeProbe("L"),
                actions:
                [
                    new IconThemeProbe("A1"),
                    new IconThemeProbe("A2"),
                ]));
    }

    private Widget BuildDefaultReferencePreview()
    {
        return new AppBar(
            titleText: "Default icon chain",
            backgroundColor: AppBarBackground,
            foregroundColor: _foregroundColor,
            leading: new IconThemeProbe("L"),
            actions:
            [
                new IconThemeProbe("A1"),
                new IconThemeProbe("A2"),
            ]);
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void SetForeground(Color value)
    {
        SetState(() => _foregroundColor = value);
    }

    private void SetThemeIconEnabled(bool value)
    {
        SetState(() => _themeIconEnabled = value);
    }

    private void SetThemeActionsIconEnabled(bool value)
    {
        SetState(() => _themeActionsIconEnabled = value);
    }

    private void SetWidgetIconOverrideEnabled(bool value)
    {
        SetState(() => _widgetIconOverrideEnabled = value);
    }

    private void SetWidgetActionsIconOverrideEnabled(bool value)
    {
        SetState(() => _widgetActionsIconOverrideEnabled = value);
    }

    private Color ResolveExpectedLeadingColor()
    {
        if (_widgetIconOverrideEnabled)
        {
            return WidgetIconColor;
        }

        if (_themeIconEnabled)
        {
            return ThemeIconColor;
        }

        return _foregroundColor;
    }

    private Color ResolveExpectedActionsColor()
    {
        if (_widgetActionsIconOverrideEnabled)
        {
            return WidgetActionsIconColor;
        }

        if (_themeActionsIconEnabled)
        {
            return ThemeActionsIconColor;
        }

        if (_widgetIconOverrideEnabled)
        {
            return WidgetIconColor;
        }

        if (_themeIconEnabled)
        {
            return ThemeIconColor;
        }

        return _foregroundColor;
    }

    private static string OnOff(bool value) => value ? "on" : "off";

    private static string ColorLabel(Color color) => $"#{color.A:X2}{color.R:X2}{color.G:X2}{color.B:X2}";
}

internal sealed class IconThemeProbe : StatelessWidget
{
    private readonly string _label;

    public IconThemeProbe(string label)
    {
        _label = label;
    }

    public override Widget Build(BuildContext context)
    {
        var iconTheme = IconTheme.Of(context);
        var swatch = iconTheme.Color ?? Colors.Transparent;
        var sizeLabel = iconTheme.Size.HasValue ? iconTheme.Size.Value.ToString("0") : "-";

        return new Container(
            width: 58,
            height: 24,
            color: Color.Parse("#FFF4F7FB"),
            decoration: new BoxDecoration(
                Border: new BorderSide(Color.Parse("#FFB8C4D4"), 1),
                BorderRadius: BorderRadius.Circular(6)),
            padding: new Thickness(4, 2, 4, 2),
            child: new Row(
                spacing: 3,
                children:
                [
                    new Container(
                        width: 10,
                        height: 10,
                        color: swatch),
                    new Text($"{_label}:{sizeLabel}", fontSize: 9, color: Colors.Black),
                ]));
    }
}

File: dart_sample/lib/app_bar_icon_theme_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class AppBarIconThemeDemoPage extends StatefulWidget {
  const AppBarIconThemeDemoPage({super.key});

  @override
  State<AppBarIconThemeDemoPage> createState() =>
      _AppBarIconThemeDemoPageState();
}

class _AppBarIconThemeDemoPageState extends State<AppBarIconThemeDemoPage> {
  static const Color _appBarBackground = Color(0xFF1E3A5F);
  static const Color _foregroundWhite = Colors.white;
  static const Color _foregroundMint = Color(0xFFB8FFF1);
  static const Color _themeIconColor = Color(0xFF2A9D8F);
  static const Color _themeActionsIconColor = Color(0xFFF4A261);
  static const Color _widgetIconColor = Color(0xFF8E44AD);
  static const Color _widgetActionsIconColor = Color(0xFFE63946);

  Color _foregroundColor = _foregroundWhite;
  bool _themeIconEnabled = true;
  bool _themeActionsIconEnabled = false;
  bool _widgetIconOverrideEnabled = false;
  bool _widgetActionsIconOverrideEnabled = false;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AppBar iconTheme/actionsIconTheme',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Runtime probe for icon-theme precedence: widget override > appBarTheme > iconTheme fallback > foreground fallback.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'fg white',
              onTap: () => _setForeground(_foregroundWhite),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'fg mint',
              onTap: () => _setForeground(_foregroundMint),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme icon off',
              onTap: () => _setThemeIconEnabled(false),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'theme icon on',
              onTap: () => _setThemeIconEnabled(true),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme actions off',
              onTap: () => _setThemeActionsIconEnabled(false),
              width: 118,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'theme actions on',
              onTap: () => _setThemeActionsIconEnabled(true),
              width: 118,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget icon off',
              onTap: () => _setWidgetIconOverrideEnabled(false),
              width: 114,
              background: const Color(0xFFE4ECF7),
            ),
            _buildButton(
              label: 'widget icon on',
              onTap: () => _setWidgetIconOverrideEnabled(true),
              width: 114,
              background: const Color(0xFFE4ECF7),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget actions off',
              onTap: () => _setWidgetActionsIconOverrideEnabled(false),
              width: 126,
              background: const Color(0xFFF5E8ED),
            ),
            _buildButton(
              label: 'widget actions on',
              onTap: () => _setWidgetActionsIconOverrideEnabled(true),
              width: 126,
              background: const Color(0xFFF5E8ED),
            ),
          ],
        ),
        Text(
          'fg=${_colorLabel(_foregroundColor)}, themeIcon=${_onOff(_themeIconEnabled)}, themeActions=${_onOff(_themeActionsIconEnabled)}, widgetIcon=${_onOff(_widgetIconOverrideEnabled)}, widgetActions=${_onOff(_widgetActionsIconOverrideEnabled)}, expectedLeading=${_colorLabel(_resolveExpectedLeadingColor())}, expectedActions=${_colorLabel(_resolveExpectedActionsColor())}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 320,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(10),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Themed app bar preview',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildThemedPreview(),
              const Text(
                'Default app bar reference (no icon theme overrides)',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildDefaultReferencePreview(),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildThemedPreview() {
    return Theme(
      data: ThemeData.light().copyWith(
        appBarTheme: AppBarTheme(
          backgroundColor: _appBarBackground,
          foregroundColor: _foregroundColor,
          iconTheme: _themeIconEnabled
              ? const IconThemeData(color: _themeIconColor, size: 18)
              : null,
          actionsIconTheme: _themeActionsIconEnabled
              ? const IconThemeData(color: _themeActionsIconColor, size: 16)
              : null,
        ),
      ),
      child: AppBar(
        title: const Text('Theme icon chain'),
        iconTheme: _widgetIconOverrideEnabled
            ? const IconThemeData(color: _widgetIconColor, size: 20)
            : null,
        actionsIconTheme: _widgetActionsIconOverrideEnabled
            ? const IconThemeData(color: _widgetActionsIconColor, size: 22)
            : null,
        leading: const _IconThemeProbe(label: 'L'),
        actions: const <Widget>[
          _IconThemeProbe(label: 'A1'),
          _IconThemeProbe(label: 'A2'),
        ],
      ),
    );
  }

  Widget _buildDefaultReferencePreview() {
    return AppBar(
      title: const Text('Default icon chain'),
      backgroundColor: _appBarBackground,
      foregroundColor: _foregroundColor,
      leading: const _IconThemeProbe(label: 'L'),
      actions: const <Widget>[
        _IconThemeProbe(label: 'A1'),
        _IconThemeProbe(label: 'A2'),
      ],
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _setForeground(Color value) {
    setState(() {
      _foregroundColor = value;
    });
  }

  void _setThemeIconEnabled(bool value) {
    setState(() {
      _themeIconEnabled = value;
    });
  }

  void _setThemeActionsIconEnabled(bool value) {
    setState(() {
      _themeActionsIconEnabled = value;
    });
  }

  void _setWidgetIconOverrideEnabled(bool value) {
    setState(() {
      _widgetIconOverrideEnabled = value;
    });
  }

  void _setWidgetActionsIconOverrideEnabled(bool value) {
    setState(() {
      _widgetActionsIconOverrideEnabled = value;
    });
  }

  Color _resolveExpectedLeadingColor() {
    if (_widgetIconOverrideEnabled) {
      return _widgetIconColor;
    }

    if (_themeIconEnabled) {
      return _themeIconColor;
    }

    return _foregroundColor;
  }

  Color _resolveExpectedActionsColor() {
    if (_widgetActionsIconOverrideEnabled) {
      return _widgetActionsIconColor;
    }

    if (_themeActionsIconEnabled) {
      return _themeActionsIconColor;
    }

    if (_widgetIconOverrideEnabled) {
      return _widgetIconColor;
    }

    if (_themeIconEnabled) {
      return _themeIconColor;
    }

    return _foregroundColor;
  }

  static String _onOff(bool value) => value ? 'on' : 'off';

  static String _colorLabel(Color color) {
    final int value = color.toARGB32();
    return '#${value.toRadixString(16).toUpperCase().padLeft(8, '0')}';
  }
}

class _IconThemeProbe extends StatelessWidget {
  const _IconThemeProbe({required this.label});

  final String label;

  @override
  Widget build(BuildContext context) {
    final IconThemeData theme = IconTheme.of(context);
    final Color swatch = theme.color ?? Colors.transparent;
    final String sizeLabel = theme.size == null
        ? '-'
        : theme.size!.toStringAsFixed(0);

    return Container(
      width: 58,
      height: 24,
      decoration: BoxDecoration(
        color: const Color(0xFFF4F7FB),
        border: Border.all(color: const Color(0xFFB8C4D4), width: 1),
        borderRadius: BorderRadius.circular(6),
      ),
      padding: const EdgeInsets.fromLTRB(4, 2, 4, 2),
      child: Row(
        spacing: 3,
        children: <Widget>[
          Container(width: 10, height: 10, color: swatch),
          Text(
            '$label:$sizeLabel',
            style: const TextStyle(fontSize: 9, color: Colors.black),
          ),
        ],
      ),
    );
  }
}
